proxy/src/transport/tls/conditional_accept.rs

```diff
@@ -0,0 +1,248 @@
+
+use super::{Identity, untrusted};
+
+#[derive(Debug, Eq, PartialEq)]
+pub enum Match {
+    Incomplete,
+    Matched,
+    NotMatched,
+}
+
+/// Determintes whether the given `input` looks like the start of a TLS
+/// connection that the proxy should terminate.
+///
+/// The determination is made based on whether the input looks like (the start
+/// of) a valid ClientHello that a reasonable TLS client might send, and the
+/// SNI matches the given identity.
+///
+/// XXX: Once the TLS record header is matched, the determination won't be
+/// made until the entire TLS record including the entire ClientHello handshake
+/// message is available. TODO: Reject non-matching inputs earlier.
+///
+/// This assumes that the ClientHello is small and is sent in a single TLS
+/// record, which is what all reasonable implementations do. (If they were not
+/// to, they wouldn't interoperate with picky servers.)
+pub fn match_client_hello(input: &[u8], identity: &Identity) -> Match {
+    let r = untrusted::Input::from(input).read_all(untrusted::EndOfInput, |input| {
+        let r = extract_sni(input);
+        input.skip_to_end(); // Ignore anything after what we parsed.
+        r
+    });
+    match r {
+        Ok(Some(sni)) => {
+            let matches = if let Ok(sni) = Identity::from_sni_hostname(sni.as_slice_less_safe()) {
+                if sni == *identity {
+                    Match::Matched
+                } else {
+                    Match::NotMatched
+                }
+            } else {
+                Match::NotMatched
+            };
+            trace!("match_client_hello: parsed correctly up to SNI; matches: {:?}", matches);
+            matches
+        },
+        Ok(None) => {
+            trace!("match_client_hello: failed to parse up to SNI");
+            Match::NotMatched
+        },
+        Err(untrusted::EndOfInput) => {
+            trace!("match_client_hello: needs more input");
+            Match::Incomplete
+        },
+    }
+}
+
+/// The result is `Ok(Some(hostname))` if the SNI extension was found, `Ok(None)`
+/// if we affirmatively rejected the input before we found the SNI extension, or
+/// `Err(EndOfInput)` if we don't have enough input to continue.
+fn extract_sni<'a>(input: &mut untrusted::Reader<'a>)
+    -> Result<Option<untrusted::Input<'a>>, untrusted::EndOfInput>
+{
+    // TLS ciphertext record header.
+
+    if input.read_byte()? != 22 { // ContentType::handshake
+        return Ok(None);
+    }
+    if input.read_byte()? != 0x03 { // legacy_record_version.major is always 0x03.
+        return Ok(None);
+    }
+    {
+        // legacy_record_version.minor may be 0x01 or 0x03 according to
+        // https://tools.ietf.org/html/draft-ietf-tls-tls13-28#section-5.1
+        let minor = input.read_byte()?;
+        if minor != 0x01 && minor != 0x03 {
+            return Ok(None);
+        }
+    }
+
+    // Treat the record length and its body as a vector<u16>.
+    let r = read_vector(input, |input| {
```

Comment:
TIOLI: This parser is fairly understandable (and the comments make it more so!) but I wonder if giving some of these closures names by assigning them to `let` bindings (or just making them local functions) would make it even clearer? Just a thought; perhaps it would only complicate things.

Reply:
I think there's definitely stuff that could be done to make this clearer but I don't think reorganizing it like that is going to have a large enough benefit to justify any more delay. In the future we'll probably extend this parser to do more interesting things; I suggest we consider refactoring it at that point. WDYT?

Reply:
👍